docs/man/ohhandler.pod.1
=head1 NAME

ohhandler - An openhpi sample application that uses the openhpi additional interfaces to control openhpi configuration. That is plugins can be loaded, plugin instances (handlers) created, unloaded or their configuration displayed.

=head1 SYNOPSIS 

 ohhandler [-D nn] [-N host[:port]] [-C <cfgfile>] [-X] command <specific arguments>
 ohhandler [--domain nn] [--host=host[:port]] [--cfgfile=file] [--debug] command <specific arguments>

 ohhandler [-D nn] [-x] list
 ohhandler [-D nn] [-x] info    <handler-id>
 ohhandler [-D nn] [-x] destroy <handler-id>
 ohhandler [-D nn] [-x] getnext <handler-id>
 ohhandler [-D nn] [-x] find    <resource-id>
 ohhandler [-D nn] [-x] retry   <handler-id>
 ohhandler [-D nn] [-x] create  plugin <plugin-name> <configuration-parameters>

=head1 DESCRIPTION

ohhandler list will display a list of all plugin instances (handlers) currently defined in the openhpi daemon for the specified domain. Part of the output is the handler-id that can be used in subsequent commands.

ohhandler info will display specific info for the handler with the specified id. This information includes all configuration parameters of the handler as specified in the openhpi.conf file or as dynamically defined during a "ohhandler create" call.

ohhandler destroy will unload the specified handler and delete all its configuration information.

ohhandler getnext will allows to walk through all hanlers currently defined in the openhpi daemon for the specified domain.

ohhandler find will find the handler that is managing the specified resource.

ohhandler retry allows to try again to load and initialize the specified handler.

ohhandler create allows to dynamically create a new handler with configuration parameters like they are specified in the openhpi.conf file. 
 - The type of plugin is specified with the keyword plugin
 - Configuration parameters should follow as name value pairs
 - Complex strings like entity paths must be enclosed with ""
 Example:
 ohhandler create plugin libsimulator entity_root "{SYSTEM_CHASSIS,1}" name sim

If no domain or host is selected, ohhandler uses the default domain as specified in the openhpiclient.conf file.

=head1 OPTIONS

=over 2

Help Options:
  -h, --help                   Show help options

Application Options:
  -D, --domain=nn              Select domain id nn
  -X, --debug                  Display debug messages
  -N, --host="host<:port>"     Open session to the domain served by the daemon 
                               at the specified URL (host:port)
                               This option overrides the OPENHPI_DAEMON_HOST and
                               OPENHPI_DAEMON_PORT environment variables.
  -C, --cfgfile="file"         Use passed file as client configuration file
                               This option overrides the OPENHPICLIENT_CONf
                               environment variable.


=back

=head1 SEE ALSO

         hpi_shell

         hpialarms      hpifan         hpipower       hpithres
         hpidomain      hpigensimdata  hpireset       hpitop
         hpiel          hpiiinv        hpisensor      hpitree
         hpievents      hpionIBMblade  hpisettime     hpiwdt
         hpixml
         ohdomainlist   ohparam

 
=head1 AUTHORS

Authors of this man page:

 Ulrich Kleber (ulikleber@users.sourceforge.net)



clients/ohhandler.c
/*      -*- linux-c -*-
 *
 * Copyright (C) Copyright Nokia Siemens Networks 2010
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  This
 * file and program are licensed under a BSD style license.  See
 * the Copying file included with the OpenHPI distribution for
 * full licensing terms.
 *
 * Authors:
 *     Ulrich Kleber <ulikleber@users.sourceforge.net>
 *
 * Log: 
 *     Start from hpidomain.c 
 *     This routine controls the handlers loaded by OpenHPI daemons
 *     using the OpenHPI extensions as described in ohpi.c 
 *
 * Changes:
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <uuid/uuid.h>
#include <SaHpi.h> 
#include <oh_utils.h>
#include <oh_clients.h>
#include <oHpi.h>

#define OH_SVN_REV "$Revision: 7133 $"

/********************************************/ 
/* Function prototypes                      */
/********************************************/
static SaErrorT exechandlercreate(int, char **, int);
static SaErrorT exechandlerdestroy(oHpiHandlerIdT);
static SaErrorT exechandlerinfo(oHpiHandlerIdT);
static SaErrorT exechandlergetnext(oHpiHandlerIdT);
static SaErrorT exechandlerfind(SaHpiResourceIdT);
static SaErrorT exechandlerretry(oHpiHandlerIdT);
static SaErrorT exechandlerlist(void);


/********************************************/ 
/* Globals                                  */
/********************************************/
int fdebug     = 0;
SaHpiDomainIdT domainid = SAHPI_UNSPECIFIED_DOMAIN_ID;
SaHpiSessionIdT sessionid;


/********************************************/ 
/* Main                                     */
/********************************************/
int
main(int argc, char **argv)
{
   SaErrorT    rv = SA_OK;
   
   oHpiHandlerIdT handlerid = 0;
   SaHpiResourceIdT resid = 0;
   SaHpiBoolT printusage = FALSE;
   int i=1;

   enum cmdT {eUndefined,
      eHandlerCreate, 
      eHandlerDestroy, 
      eHandlerInfo, 
      eHandlerGetNext, 
      eHandlerFind,
      eHandlerRetry,
      eHandlerList} cmd=eUndefined;
       
   oh_prog_version(argv[0], OH_SVN_REV);

   while (i<argc && !printusage && cmd!=eHandlerCreate) {
      if (strcmp(argv[i],"-D")==0) {
         if (++i<argc) domainid = atoi(argv[i]);
         else printusage = TRUE;
      }
      else if (strcmp(argv[i],"-x")==0) fdebug=1;

      else if (strcmp(argv[i],"create")==0) {
         cmd=eHandlerCreate;
         // exechandlercreate will do the remaining reading of 
         // parameters itself 
         rv = exechandlercreate (argc, argv, ++i);
         if (rv == SA_OK) exit (0);
         if (rv != SA_ERR_HPI_INVALID_PARAMS) exit (1);
         printusage = TRUE;
      }

      else if (strcmp(argv[i],"destroy")==0) {
         cmd=eHandlerDestroy;
         if (++i<argc) handlerid = atoi(argv[i]);
         else printusage = TRUE;
      }

      else if (strcmp(argv[i],"info")==0) {
         cmd=eHandlerInfo;
         if (++i<argc) handlerid = atoi(argv[i]);
         else printusage = TRUE;
      }

      else if (strcmp(argv[i],"getnext")==0) {
         cmd=eHandlerGetNext;
         if (++i<argc) handlerid = atoi(argv[i]);
         else printusage = TRUE;
      }

      else if (strcmp(argv[i],"find")==0) {
         cmd=eHandlerFind;
         if (++i<argc) resid = atoi(argv[i]);
         else printusage = TRUE;
      }

      else if (strcmp(argv[i],"retry")==0) {
         cmd=eHandlerRetry;
         if (++i<argc) handlerid = atoi(argv[i]);
         else printusage = TRUE;
      }

      else if (strcmp(argv[i],"list")==0) {
         cmd=eHandlerList;
         if (++i<argc) printusage = TRUE;
      }

      else printusage = TRUE;

      i++;
   }

   if (cmd == eHandlerCreate) {
      rv = exechandlercreate (argc, argv, i);
      if (rv == SA_OK) exit (0);
      if (rv == SA_ERR_HPI_INVALID_PARAMS)
         printusage = TRUE;
      else exit (1);
   }

   if (printusage == TRUE || cmd == eUndefined)
   {
      printf("\n");
      printf("Usage: %s [-D domain] [-x] command [specific arguments]\n\n",
         argv[0]);
      printf("      -D nn  Select domain id nn (not supported yet by oh-functions)\n");
      printf("      -x     Display debug messages\n");
      printf("\n");
      printf("    ohhander [-D domain] [-x] list\n");
      printf("             List the handlers loaded in specified domain\n\n");
      printf("    ohhander [-D domain] [-x] info <handler-id>\n");
      printf("             Display info about handler <handler-id> \n\n");
      printf("    ohhander [-D domain] [-x] destroy <handler-id>\n");
      printf("             Unload handler <handler-id> and delete its config \n\n");
      printf("    ohhander [-D domain] [-x] getnext <handler-id>\n");
      printf("             Find next valid handler id from <handler-id> \n\n");
      printf("    ohhander [-D domain] [-x] find <resource-id>\n");
      printf("             Find the right handler for a resource id \n\n");
      printf("    ohhander [-D domain] [-x] retry <handler-id>\n");
      printf("             Retry loading of handler <handler-id> \n\n");
      printf("    ohhander [-D domain] [-x] create plugin <name> <params>\n");
      printf("             Create handler with the specified parameters.\n");
      printf("             Pairs of strings in commandline like in openhpi.conf.\n");
      printf("             Keyword plugin to select type of handler.\n");
      printf("             Entity root and other complex strings must be \n");
      printf("             enclosed in \".\n");
      printf("     example:\n");
      printf("             ohhandler create plugin libsimulator "
             "entity_root \"{SYSTEM_CHASSIS,1}\" name sim\n");

      printf("\n\n");
      exit(1);
   }

   if (fdebug) {
      if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
         printf("saHpiSessionOpen\n");
      else printf("saHpiSessionOpen to domain %u\n",domainid);
   }
        rv = saHpiSessionOpen(domainid,&sessionid,NULL);
   if (rv != SA_OK) {
      printf("saHpiSessionOpen returns %d (%s)\n",
         rv, oh_lookup_error(rv));
      exit(-1);
   }
   if (fdebug)
             printf("saHpiSessionOpen returns with SessionId %u\n", 
         sessionid);

   switch (cmd){
                case eHandlerCreate: break; //already done
                case eHandlerDestroy:
                                   rv = exechandlerdestroy ( handlerid );
                                   break;
                case eHandlerInfo:
                                   rv = exechandlerinfo ( handlerid );
                                   break;
                case eHandlerGetNext:
                                   rv = exechandlergetnext ( handlerid );
                                   break;
                case eHandlerFind:
                                   rv = exechandlerfind ( resid );
                                   break;
                case eHandlerRetry:
                                   rv = exechandlerretry ( handlerid );
                                   break;
                case eHandlerList:
                                   rv = exechandlerlist ( );
                                   break;
                case eUndefined:   break; //already done
   }

   if (fdebug) {
      printf("Internal processing returns %d (%s)\n",
         rv, oh_lookup_error(rv));
      printf("Calling saHpiSessionClose for session %u\n",sessionid);
   }

   rv = saHpiSessionClose(sessionid);
   if (fdebug) {
       printf("saHpiSessionClose returns %d (%s)\n",
          rv, oh_lookup_error(rv));
   }     
       
   exit(0);
}

/********************************************/ 
/* exechandlercreate                        */
/********************************************/
static SaErrorT exechandlercreate (int argc, char **argv, int i)
{
   SaErrorT rv = SA_OK;
   oHpiHandlerIdT handlerid = 0;
   GHashTable * createparams = g_hash_table_new_full (
      g_str_hash, g_str_equal, g_free, g_free);
   SaHpiBoolT pluginnamegiven = SAHPI_FALSE;

   if (fdebug) printf ("createhandler started\n");

   while (i<argc){
      if (strcmp(argv[i],"-f")==0) {
         printf("input from file not implemented yet\n");
         return (SA_OK);
      }
      else if (++i<argc) {
         if (strcmp(argv[i-1],"plugin")==0) pluginnamegiven = SAHPI_TRUE;
         g_hash_table_insert( createparams,
            g_strdup( argv[i-1] ), 
            g_strdup( argv[i] ));
          if (fdebug) printf ("Pair of arguments: %s - %s\n",
            g_strdup( argv[i-1] ), 
            g_strdup( argv[i] ));
      }
      else // parameters not in pairs
         return (SA_ERR_HPI_INVALID_PARAMS);
      i++;
   }

   if (!pluginnamegiven) {
      printf("You must enter a valid plugin name\n");
      return (SA_ERR_HPI_INVALID_PARAMS);
   }

   if (fdebug) {
      if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID)
         printf("saHpiSessionOpen\n");
      else printf("saHpiSessionOpen to domain %u\n",domainid);
   }
   rv = saHpiSessionOpen(domainid,&sessionid,NULL);
   if (rv != SA_OK) {
      printf("saHpiSessionOpen returns %d (%s)\n",
             rv, oh_lookup_error(rv));
      return (SA_ERR_HPI_INVALID_PARAMS);
   }
   if (fdebug)
      printf("saHpiSessionOpen returns with SessionId %u\n",
             sessionid);

   if (fdebug) printf ("Calling oHpiHandlerCreate!\n");
   rv = oHpiHandlerCreate(sessionid, createparams, &handlerid );

   if ( rv != SA_OK ) {
      printf("oHpiHandlerCreate returned %d (%s)\n",
         rv, oh_lookup_error(rv));
      return(rv);
   }
   
   printf("Handler %u successfully created!\n", handlerid);

   rv = saHpiSessionClose(sessionid);
        
   return(SA_OK);

}

/********************************************/ 
/* exechandlerdestroy                       */
/********************************************/
static SaErrorT exechandlerdestroy(oHpiHandlerIdT handlerid)
{
   SaErrorT rv = SA_OK;
   if (fdebug) printf("Go and unload handler %u in domain %u\n", 
            handlerid, domainid);
   
   rv = oHpiHandlerDestroy ( sessionid, handlerid );

   if (fdebug) printf("oHpiHandlerDestroy returned %d (%s)\n", 
         rv, oh_lookup_error(rv));

   if (rv==SA_OK) {
      printf("Handler %u successfully unloaded.\n\n", handlerid);
      return rv;
   }
   else if (rv==SA_ERR_HPI_NOT_PRESENT) {
      if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
         printf("Handler %u is not existing in default domain.\n",
                handlerid);
      else printf("Handler %u is not existing in domain %u.\n",
                handlerid, domainid);
      return rv;
   }
   else printf("\nHandler %u couldn't be unloaded, Returncode %d (%s)\n", 
         handlerid, rv, oh_lookup_error(rv));
   return rv;
}

/********************************************/ 
/* exechandlerinfo                          */
/********************************************/
// function to print from hash table
static void print_pair (gpointer key, gpointer value, gpointer user_data)
{
   // we already printed the pluginname and entity-root
   if (strcmp((gchar *)key, "plugin")==0) return;
   if (strcmp((gchar *)key, "entity_root")==0) return;
   printf ("   %s %s\n", (gchar *)key, (gchar *)value);
}

static 
SaErrorT exechandlerinfo(oHpiHandlerIdT handlerid)
{
   oHpiHandlerInfoT handlerinfo;
   SaErrorT rv = SA_OK;
   GHashTable *handlerconfig = g_hash_table_new_full(
                        g_str_hash, g_str_equal,
                        g_free, g_free );

   if (fdebug) printf("Go and display handler info for %u\n", handlerid);
   
   rv = oHpiHandlerInfo ( sessionid, handlerid, &handlerinfo, handlerconfig );

   if (rv==SA_ERR_HPI_NOT_PRESENT) {
      if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
         printf("Handler %u is not existing in default domain.\n",
                handlerid);
      else printf("Handler %u is not existing in domain %u.\n",
                handlerid, domainid);
      g_hash_table_destroy(handlerconfig);
      return SA_OK;
   }
   else if (rv!=SA_OK) {
      printf("oHpiHandlerInfo returned %d (%s)\n", 
                rv, oh_lookup_error(rv));
      g_hash_table_destroy(handlerconfig);
      return rv;
   }
   if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
      printf("\n\nInfo for handler %u in default domain:\n\n",handlerid);
   else printf("\n\nInfo for handler %u in domain %u:\n\n",handlerid, domainid);
   printf("Plugin name: %s\n",handlerinfo.plugin_name);

   printf("Root ");
   oh_print_ep (&(handlerinfo.entity_root),0);

   printf("Failed attempts to load handler: %u\n",handlerinfo.load_failed);

   printf("\nHandler configuration:\n");
   printf ("   plugin %s\n", handlerinfo.plugin_name);
   printf ("   entity_root \"%s\"\n",(const char *) g_hash_table_lookup(handlerconfig, "entity_root")); 
   g_hash_table_foreach(handlerconfig, print_pair, NULL);
   printf("\n\n");

   g_hash_table_destroy(handlerconfig);
   return rv;
}

/********************************************/ 
/* exechandlergetnext                       */
/********************************************/
static SaErrorT exechandlergetnext(oHpiHandlerIdT handlerid)
{
   SaErrorT rv = SA_OK;
   oHpiHandlerIdT nexthandlerid;

   if (fdebug) printf("Go and get next handler from %u in domain %u\n",
                      handlerid, domainid);

   rv = oHpiHandlerGetNext ( sessionid, handlerid, &nexthandlerid );

   if (fdebug) printf("oHpiHandlerGetNext returned %d (%s)\n",
                      rv, oh_lookup_error(rv));

   if (rv==SA_OK) {
      printf("Next Handler found from %u is %u.\n\n",
             handlerid, nexthandlerid);
   }
   else if (rv==SA_ERR_HPI_NOT_PRESENT) {
      if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
         printf("No next Handler found from %u in default domain.\n",
                        handlerid);
      else printf("No next Handler found from %u in domain %u.\n",
                        handlerid, domainid);
   }
   else printf("\noHpiHandlerGetNext returned %d (%s)\n",
               rv, oh_lookup_error(rv));
   return rv;
}

/********************************************/ 
/* exechandlerfind                          */
/********************************************/
static SaErrorT exechandlerfind(SaHpiResourceIdT resid)
{
   SaErrorT rv = SA_OK;
   oHpiHandlerIdT handlerid;

   if (fdebug) printf("Go and find handler for resource %u in domain "
                      "%u using session %u\n",
                      resid, domainid, sessionid);

   rv = oHpiHandlerFind ( sessionid, resid, &handlerid );

   if (fdebug) printf("oHpiHandlerFind returned %d (%s)\n",
                      rv, oh_lookup_error(rv));

   if (rv==SA_OK) {
      printf("Handler found for resource %u:   Handler-id %u.\n\n",
             resid, handlerid);
   }
   else if (rv==SA_ERR_HPI_NOT_PRESENT) {
      if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
         printf("No Handler found for resource %u in default domain.\n",
                      resid);
      else printf("No Handler found for resource %u in domain %u.\n",
                      resid, domainid);
   }
   else printf("\noHpiHandlerFind returned %d (%s)\n",
               rv, oh_lookup_error(rv));
   return rv;
}

/********************************************/ 
/* exechandlerretry                         */
/********************************************/
static SaErrorT exechandlerretry(oHpiHandlerIdT handlerid)
{
   SaErrorT rv = SA_OK;

   if (fdebug) printf("Go and retry loading handler %u in domain %u\n", 
                      handlerid, domainid);

   rv = oHpiHandlerRetry ( sessionid, handlerid );

   if (rv!=SA_OK) {
      printf("oHpiHandlerRetry returned %d (%s)\n",
                        rv, oh_lookup_error(rv));
      return rv;
   }
   if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
      printf("\nHandler %u in default domain successfully re-loaded.\n",
                        handlerid);
   else printf("\nHandler %u in domain %u successfully re-loaded.\n",
                        handlerid, domainid);
   return rv;
}

/********************************************/ 
/* exechandlerlist                          */
/********************************************/
static SaErrorT exechandlerlist()
{
   SaErrorT rv = SA_OK;
   oHpiHandlerIdT handlerid = 0;
   oHpiHandlerIdT nexthandlerid;
   oHpiHandlerInfoT handlerinfo;
   GHashTable *config = 0;

   if (domainid==SAHPI_UNSPECIFIED_DOMAIN_ID) 
      printf("Handlers defined in default domain:\n");
   else printf("Handlers defined in Domain %u:\n",domainid);
   while (rv==SA_OK) {
      rv = oHpiHandlerGetNext ( sessionid, handlerid, &nexthandlerid );

      if (fdebug) printf("oHpiHandlerGetNext (%u) returned %d (%s)\n",
                      handlerid, rv, oh_lookup_error(rv));

      if (rv==SA_OK) {
         //display info for that handler
         config = g_hash_table_new_full(
                        g_str_hash, g_str_equal,
                        g_free, g_free );
         rv = oHpiHandlerInfo ( sessionid, nexthandlerid, &handlerinfo, config );
         if (rv!=SA_OK) {
            printf("oHpiHandlerInfo for handler %u returned %d (%s)\n", 
                   nexthandlerid, rv, oh_lookup_error(rv));
            rv = SA_OK;
         }
         else {
            printf("Handler %u: %s, ",
               nexthandlerid, handlerinfo.plugin_name);
            oh_print_ep (&(handlerinfo.entity_root),0);
         }
         g_hash_table_destroy(config);
      }
      else if (rv!=SA_ERR_HPI_NOT_PRESENT) {
         printf("Error: oHpiHandlerGetNext (%u) returned %d (%s)\n",
                        handlerid, rv, oh_lookup_error(rv));
      }
      handlerid = nexthandlerid;
   }
   printf("\n");
   if (rv==SA_ERR_HPI_NOT_PRESENT) return SA_OK;
   return rv;
}

/* end ohhandler.c */
